### PR TITLE
Improvements to discriminated union docs

### PR DESCRIPTION
**Pull request by samuelcolvin**
* Improve discriminated union docs
* Minor improvements to `JsonValue` docs as well (removing `assert` from examples, using `BaseModel` instead of `TypeAdapter`)

### CODE REVIEW

File: docs/concepts/unions.md [hunk 1]
```diff
@@ -285,10 +288,18 @@ class DiscriminatedModel(BaseModel):
 
 data = {'x': {'x': {'x': 'a'}}}
 m = DiscriminatedModel.model_validate(data)
-assert m == (
-    DiscriminatedModel(x=DiscriminatedModel(x=DiscriminatedModel(x='a')))
-)
-assert m.model_dump() == data
+print(m)
```
Comment (samuelcolvin) [inline docs/concepts/unions.md:291]: I believe it's most reader friendly to use print than assert, can we check were else we're using `assert`.

File: docs/concepts/unions.md [hunk 2]
```diff
@@ -285,10 +288,18 @@ class DiscriminatedModel(BaseModel):
...
+#> x=DiscriminatedModel(x=DiscriminatedModel(x='a'))
+
+try:
```
Comment (samuelcolvin) [inline docs/concepts/unions.md:294]: this example, together with an example or the errors raised without a custom error need to be in a separate example.

File: pydantic/types.py [hunk 3]
```diff
@@ -2594,15 +2596,16 @@ class ThanksgivingDinner(BaseModel):
...
     ```
-
-    See the [Discriminated Unions](../concepts/unions.md#discriminated-unions)
-    docs for more details on how to use `CallableDiscriminator`s.
     """
 
     discriminator: Callable[[Any], Hashable]
+    """The callable to use to extract the value of the discriminator from the input."""
     custom_error_type: str | None = None
+    """TType to use in custom errors that replace standard errors related to the discrimination."""
```
Comment (samuelcolvin) [inline pydantic/types.py:2604]:
```suggestion
    """Type to use in custom errors that replace standard errors related to the discrimination."""
```

Comment (samuelcolvin) [reply pydantic/types.py:2604]: we need a section about custom errors to link to here.

File: docs/concepts/unions.md (hunk 1 above)
Comment (sydney-runkle) [reply docs/concepts/unions.md:292]: Agreed. I'll change the rest of the `assert` statements to `print` statements.

File: docs/concepts/unions.md [hunk 4]
```diff
@@ -259,39 +262,57 @@ from typing import Any, Union
...
+    print(e)  # (1)!
+    """
+    1 validation error for DiscriminatedModel
+    value
+      Unable to extract tag using discriminator model_x_discriminator() [type=union_tag_not_found, input_value='not an int or a model', input_type=str]
+    """
 ```
 
+1. Notice the callable discriminator function returns `None` if a discriminator value is not found.
+   When `None` is returned, this error is raised.
```
Comment (dmontagu) [inline docs/concepts/unions.md:314]:
```suggestion
   When `None` is returned, this `union_tag_not_found` error is raised.
```

File: docs/concepts/unions.md [hunk 5]
```diff
@@ -383,8 +404,11 @@ except ValidationError as e:
 
 ## Union Validation Errors
 
-When validation fails, error messages can be quite verbose, especially when you're not using discriminated unions.
-The below example shows the benefits of using discriminated unions in terms of error message simplicity.
+When `Union` validation fails, error messages can be quite verbose, especially when dealing with recursive models.
+Discriminated unions help to simplify error messages in this case.
```
Comment (dmontagu) [inline docs/concepts/unions.md:408]:
```suggestion
When `Union` validation fails, error messages can be quite verbose, as they will produce validation errors for
each case in the union.
This is especially noticeable when dealing with recursive models, where reasons may be generated at each level of
recursion.
Discriminated unions help to simplify error messages in this case, as validation errors are only produced for
the case with a matching discriminator value.
```

File: docs/concepts/unions.md [hunk 6]
```diff
@@ -401,64 +425,35 @@ class Model(BaseModel):
...
+    x.Model.x.str
+      Input should be a valid string [type=string_type, input_value={'x': {}}, input_type=dict]
+    x.Model.x.Model.x.str
+      Input should be a valid string [type=string_type, input_value={}, input_type=dict]
+    x.Model.x.Model.x.Model.x
+      Field required [type=missing, input_value={}, input_type=dict]
+    """
 
 
 # Errors are much more simple with a discriminated union:
```
Comment (dmontagu) [inline docs/concepts/unions.md:459]:
```suggestion
# Errors are much simpler with a discriminated union:
```
I know it's not currently affected by this PR but 🤷‍♂️

File: pydantic/types.py [hunk 7]
```diff
@@ -2600,9 +2602,15 @@ class ThanksgivingDinner(BaseModel):
     """
 
     discriminator: str | Callable[[Any], Hashable]
+    """The callable to use to extract the value of the discriminator from the input."""
```
Comment (dmontagu) [inline pydantic/types.py:2605]: Should probably explain what it means if it's a `str` — i.e., the name of the field to use as the discriminator, same as in the `Field` function.

File: pydantic/types.py [hunk 8]
```diff
@@ -2600,9 +2602,15 @@ class ThanksgivingDinner(BaseModel):
...
     custom_error_type: str | None = None
+    """Type to use in [custom errors](../errors/errors.md#custom-errors) that replace standard errors related to the
+    discrimination.
+    """
     custom_error_message: str | None = None
+    """Message to use in custom errors that replace standard errors related to the discrimination."""
     custom_error_context: dict[str, int | str | float] | None = None
+    """Context to use in custom errors that replace standard errors related to the discrimination."""
```
Comment (dmontagu) [inline pydantic/types.py:2613]:
```suggestion
    custom_error_type: str | None = None
    """Type to use in [custom errors](../errors/errors.md#custom-errors) replacing the standard discriminated union
    validation errors.
    """
    custom_error_message: str | None = None
    """Message to use in custom errors replacing the standard discriminated union validation errors."""
    custom_error_context: dict[str, int | str | float] | None = None
    """Context to use in custom errors replacing the standard discriminated union validation errors."""
```
or even
```suggestion
    custom_error_type: str | None = None
    """Type to use in [custom errors](../errors/errors.md#custom-errors) replacing the standard discriminated union
    validation errors.
    """
    custom_error_message: str | None = None
    """Message to use in custom errors."""
    custom_error_context: dict[str, int | str | float] | None = None
    """Context to use in custom errors."""
```

**Review by dmontagu — APPROVED** [review]
You can take or leave the suggestions above. Otherwise LGTM

File: pydantic/types.py (hunk 7 above)
Comment (sydney-runkle) [reply pydantic/types.py]: Done, thanks for catching that.

**Pull request merged**